service/history/api/recordworkflowtaskstarted/api.go

```diff
@@ -137,33 +138,45 @@ func Invoke(
 			// The stickiness info is used by frontend to decide if it should send down partial history or full history.
 			// Sending down partial history will cost the worker an extra fetch to server for the full history.
 			currentTaskQueue := mutableState.CurrentTaskQueue()
+			pollerTaskQueue := req.PollRequest.TaskQueue
 			if currentTaskQueue.Kind == enumspb.TASK_QUEUE_KIND_STICKY &&
-				currentTaskQueue.GetName() != req.PollRequest.TaskQueue.GetName() {
+				currentTaskQueue.GetName() != pollerTaskQueue.GetName() {
 				// For versioned workflows we additionally check for the poller queue to not be a sticky queue itself.
 				// Although it's ideal to check this for unversioned workflows as well, we can't rely on older clients
 				// setting the poller TQ kind.
 				if (mutableState.GetAssignedBuildId() != "" || mutableState.GetEffectiveDeployment() != nil) &&
-					req.PollRequest.TaskQueue.Kind == enumspb.TASK_QUEUE_KIND_STICKY {
+					pollerTaskQueue.Kind == enumspb.TASK_QUEUE_KIND_STICKY {
 					return nil, serviceerrors.NewObsoleteDispatchBuildId("wrong sticky queue")
 				}
 				// req.PollRequest.TaskQueue.GetName() may include partition, but we only check when sticky is enabled,
 				// and sticky queue never has partition, so it does not matter.
 				mutableState.ClearStickyTaskQueue()
 			}
 
-			deployment := worker_versioning.DeploymentFromCapabilities(req.PollRequest.WorkerVersionCapabilities)
+			if currentTaskQueue.Kind != pollerTaskQueue.Kind &&
+				// Older SDKs might not specify poller task queue kind
+				pollerTaskQueue.Kind != enumspb.TASK_QUEUE_KIND_UNSPECIFIED {
+				// Matching can drop the task, newer one should be scheduled already.
+				return nil, serviceerrors.NewObsoleteMatchingTask("wrong task queue type")
+			}
+
+			dispatchDeployment := worker_versioning.DeploymentFromCapabilities(req.PollRequest.WorkerVersionCapabilities)
+			directiveDeployment := req.GetDirectiveDeployment()
```

Comment:
```suggestion
			scheduledDeployment := req.GetScheduledDeployment()
```
-- deployment which history wanted when it scheduled a WFT.